File: python-server/storage/json_file_store.py
"""
JSON file-based storage layer.
Mirrors the Node.js jsonFileStore.ts exactly.
"""
import os
import json

DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'data')
DATA_DIR = os.path.normpath(DATA_DIR)


def _ensure_data_dir():
    os.makedirs(DATA_DIR, exist_ok=True)


def _file_path(name):
    return os.path.join(DATA_DIR, name)


def _write_json(name, data):
    _ensure_data_dir()
    with open(_file_path(name), 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)


def _read_json(name, fallback=None):
    _ensure_data_dir()
    fp = _file_path(name)
    if not os.path.exists(fp):
        return fallback
    try:
        with open(fp, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception:
        return fallback


def save_heatmap_data(data):
    _write_json('heatmap.json', data)


def load_heatmap_data():
    return _read_json('heatmap.json', [])


def save_roster_data(roster):
    _write_json('roster.json', roster)


def load_roster_data():
    return _read_json('roster.json', None)


def save_slots(slots):
    _write_json('slots.json', slots)


def load_slots():
    return _read_json('slots.json', [])


def save_session_meta(meta):
    _write_json('session.json', meta)


def load_session_meta():
    return _read_json('session.json', None)


def save_revised_heatmap(data):
    _write_json('revised_heatmap.json', data)


def load_revised_heatmap():
    return _read_json('revised_heatmap.json', [])


def save_recommendations(recs):
    _write_json('recommendations.json', recs)


def load_recommendations():
    return _read_json('recommendations.json', [])


def clear_all():
    _ensure_data_dir()
    for f in os.listdir(DATA_DIR):
        if f.endswith('.json'):
            os.unlink(os.path.join(DATA_DIR, f))


def clear_current_week():
    """Clears only current-week data, preserving historical data. Week starts on Sunday."""
    from datetime import date, timedelta
    today = date.today()
    day = today.weekday()  # Monday=0, Sunday=6
    # Python weekday: Mon=0..Sun=6. We need Sunday start.
    # days_since_sunday: Sun=0, Mon=1, Tue=2, ...
    days_since_sunday = (day + 1) % 7
    sun = today - timedelta(days=days_since_sunday)
    week_start = sun.isoformat()  # YYYY-MM-DD

    # Filter heatmap
    heatmap = load_heatmap_data()
    historical_heatmap = [r for r in heatmap if r['date'] < week_start]
    save_heatmap_data(historical_heatmap)

    # Filter roster
    roster = load_roster_data()
    if roster:
        historical_entries = [e for e in roster['entries'] if e['date'] < week_start]
        if historical_entries:
            agents = sorted(set(e['agent'] for e in historical_entries))
            managers = sorted(set(e['manager'] for e in historical_entries if e.get('manager')))
            programs = sorted(set(e['program'] for e in historical_entries if e.get('program')))
            lobbies = sorted(set(e['lobby'] for e in historical_entries if e.get('lobby')))
            dates = sorted(set(e['date'] for e in historical_entries))
            save_roster_data({
                'entries': historical_entries,
                'agents': agents,
                'managers': managers,
                'programs': programs,
                'lobbies': lobbies,
                'dates': dates,
            })
        else:
            fp = os.path.join(DATA_DIR, 'roster.json')
            if os.path.exists(fp):
                os.unlink(fp)

    # Filter slots
    slots = load_slots()
    historical_slots = [s for s in slots if s['date'] < week_start]
    save_slots(historical_slots)

    # Filter recommendations
    recs = load_recommendations()
    historical_recs = [r for r in recs if r['date'] < week_start]
    save_recommendations(historical_recs)

    # Filter revised heatmap
    revised = load_revised_heatmap()
    historical_revised = [r for r in revised if r['date'] < week_start]
    save_revised_heatmap(historical_revised)

    # Reset session meta
    fp = os.path.join(DATA_DIR, 'session.json')
    if os.path.exists(fp):
        os.unlink(fp)

File: server/src/storage/jsonFileStore.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import type { HeatmapRow, ShiftRoster, OTSlot, SessionMeta, OTRecommendation } from '../types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DATA_DIR = path.resolve(__dirname, '../../data');

function ensureDataDir(): void {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

function filePath(name: string): string {
  return path.join(DATA_DIR, name);
}

function writeJSON(name: string, data: unknown): void {
  ensureDataDir();
  fs.writeFileSync(filePath(name), JSON.stringify(data, null, 2), 'utf-8');
}

function readJSON<T>(name: string, fallback: T): T {
  ensureDataDir();
  const fp = filePath(name);
  if (!fs.existsSync(fp)) return fallback;
  try {
    const raw = fs.readFileSync(fp, 'utf-8');
    return JSON.parse(raw) as T;
  } catch {
    return fallback;
  }
}

export function saveHeatmapData(data: HeatmapRow[]): void {
  writeJSON('heatmap.json', data);
}

export function loadHeatmapData(): HeatmapRow[] {
  return readJSON<HeatmapRow[]>('heatmap.json', []);
}

export function saveRosterData(roster: ShiftRoster): void {
  writeJSON('roster.json', roster);
}

export function loadRosterData(): ShiftRoster | null {
  return readJSON<ShiftRoster | null>('roster.json', null);
}

export function saveSlots(slots: OTSlot[]): void {
  writeJSON('slots.json', slots);
}

export function loadSlots(): OTSlot[] {
  return readJSON<OTSlot[]>('slots.json', []);
}

export function saveSessionMeta(meta: SessionMeta): void {
  writeJSON('session.json', meta);
}

export function loadSessionMeta(): SessionMeta | null {
  return readJSON<SessionMeta | null>('session.json', null);
}

export function saveRevisedHeatmap(data: HeatmapRow[]): void {
  writeJSON('revised_heatmap.json', data);
}

export function loadRevisedHeatmap(): HeatmapRow[] {
  return readJSON<HeatmapRow[]>('revised_heatmap.json', []);
}

export function saveRecommendations(recs: OTRecommendation[]): void {
  writeJSON('recommendations.json', recs);
}

export function loadRecommendations(): OTRecommendation[] {
  return readJSON<OTRecommendation[]>('recommendations.json', []);
}

export function clearAll(): void {
  ensureDataDir();
  const files = fs.readdirSync(DATA_DIR);
  for (const file of files) {
    if (file.endsWith('.json')) {
      fs.unlinkSync(path.join(DATA_DIR, file));
    }
  }
}

/**
 * Clears only current-week data, preserving historical data.
 * Week starts on Sunday.
 */
export function clearCurrentWeek(): void {
  const now = new Date();
  const day = now.getDay();
  const sun = new Date(now);
  sun.setDate(sun.getDate() - day);
  const weekStart = `${sun.getFullYear()}-${String(sun.getMonth() + 1).padStart(2, '0')}-${String(sun.getDate()).padStart(2, '0')}`;

  // Filter heatmap — keep rows before current week
  const heatmap = loadHeatmapData();
  const historicalHeatmap = heatmap.filter((r) => r.date < weekStart);
  saveHeatmapData(historicalHeatmap);

  // Filter roster — keep entries before current week
  const roster = loadRosterData();
  if (roster) {
    const historicalEntries = roster.entries.filter((e) => e.date < weekStart);
    if (historicalEntries.length > 0) {
      const agents = [...new Set(historicalEntries.map((e) => e.agent))].sort();
      const managers = [...new Set(historicalEntries.filter((e) => e.manager).map((e) => e.manager))].sort();
      const programs = [...new Set(historicalEntries.filter((e) => e.program).map((e) => e.program))].sort();
      const lobbies = [...new Set(historicalEntries.filter((e) => e.lobby).map((e) => e.lobby))].sort();
      const dates = [...new Set(historicalEntries.map((e) => e.date))].sort();
      saveRosterData({ entries: historicalEntries, agents, managers, programs, lobbies, dates });
    } else {
      // No historical entries — remove roster file
      const fp = path.join(DATA_DIR, 'roster.json');
      if (fs.existsSync(fp)) fs.unlinkSync(fp);
    }
  }

  // Filter slots — keep slots before current week
  const slots = loadSlots();
  const historicalSlots = slots.filter((s) => s.date < weekStart);
  saveSlots(historicalSlots);

  // Filter recommendations — keep recs before current week
  const recs = loadRecommendations();
  const historicalRecs = recs.filter((r) => r.date < weekStart);
  saveRecommendations(historicalRecs);

  // Filter revised heatmap — keep rows before current week
  const revised = loadRevisedHeatmap();
  const historicalRevised = revised.filter((r) => r.date < weekStart);
  saveRevisedHeatmap(historicalRevised);

  // Reset session meta
  const fp = path.join(DATA_DIR, 'session.json');
  if (fs.existsSync(fp)) fs.unlinkSync(fp);
}

/** Exported for testing */
export { DATA_DIR };
